### PR TITLE
Stateless masternode

### PR DESCRIPTION
**Pull request by audebert**
With this PR masternode now queries the DB periodically to get the list of tasks and does not maintain that state in memory.

This enable DB updates that are reflected in masternode.

Known issue: this will result in heavy load on the DB if there is a large amount of pending tasks returned at each query. This will be mitigated in the future change with rate limiting.

### CODE REVIEW

File: python-lib/prologin/masternode/master.py [hunk 1]
```diff
@@ -220,64 +230,58 @@ def redispatch_timeout_tasks(self, worker):
...
-        for r in c:
-            logging.info('request match id %s launch', r[0])
-            mid = r[0]
-            map_contents = r[1]
-            players = list(zip(r[2], r[3], r[4]))
-            to_set_pending.append(
-                {'match_id': mid, 'match_status': 'pending',}
-            )
+        tasks = []
+        for row in rows:
```
Comment (Zopieux) [inline python-lib/prologin/masternode/master.py]: `for chamion_id, username in rows:`

File: python-lib/prologin/masternode/master.py [hunk 2]
```diff
@@ -220,64 +230,58 @@ def redispatch_timeout_tasks(self, worker):
...
+            t = CompilationTask(self.config, self.db, row[1], row[0])
+            tasks.append(t)
+
+        return tasks
+
+    async def get_requested_matches(self, status="new"):
+        rows = await self.db.execute("get_matches", {"match_status": status})
+
+        tasks = []
+        for row in rows:
```
Comment (Zopieux) [inline python-lib/prologin/masternode/master.py]: `for match_id, map_data, champion_ids, player_ids, usernames in rows:`

File: python-lib/prologin/masternode/master.py [hunk 3]
```diff
@@ -296,41 +300,14 @@ def find_worker_for(self, task):
...
-                continue
+    async def dispatch_tasks(self, queue_name, tasks):
+        logging.info("%d tasks in %s queue", len(tasks), queue_name)
+
+        for task in tasks:
+            w = self.find_worker_for(task)
+            if w is None:
+                logging.info("no worker available for task %s", task)
+                break
+            else:
```
Comment (Zopieux) [inline python-lib/prologin/masternode/master.py]: remove `else` and dedent

File: python-lib/prologin/masternode/task.py [hunk 4]
```diff
@@ -71,16 +72,29 @@ def match_path(config, match_id):
...
         self.executions = 0
 
-    def execute(self):
+    @property
+    @abc.abstractmethod
+    def slots_taken(self):
+        raise NotImplementedError
+
+    async def execute(self):
         self.start_time = time.time()
```
Comment (Zopieux) [inline python-lib/prologin/masternode/task.py]: `time.monotonic()`

File: python-lib/prologin/masternode/task.py [hunk 5]
```diff
@@ -71,16 +72,29 @@ def match_path(config, match_id):
...
+    async def redispatch(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    async def discard(self):
+        raise NotImplementedError
+
     def has_timeout(self):
         return (
             self.timeout is not None
```
Comment (Zopieux) [inline python-lib/prologin/masternode/task.py]: `time.monotonic()`

File: python-lib/prologin/masternode/master.py [hunk 6]
```diff
@@ -315,41 +347,26 @@ def find_worker_for(self, task):
...
+        logging.info("%d tasks in %s queue", len(tasks), queue_name)
+
+        coros = []
+        for task in tasks:
+            w = self.find_worker_for(task)
+            if w is None:
+                logging.info("no worker available for task %s", task)
+                # It's okay to break the loop because all the items in the task
+                # queue are expected to have the same requirements, hence if
+                # one task cannot be scheduled then none of the other will.
```
Comment (Zopieux) [inline python-lib/prologin/masternode/master.py]: others*

File: python-lib/prologin/masternode/task.py [hunk 7]
```diff
@@ -123,15 +163,43 @@ def __init__(self, config, mid, players, map_contents):
...
-        await worker.rpc.run_match(self.mid, self.players, self.map_contents)
+        try:
+            await worker.rpc.run_match(
+                self.mid, self.players, self.map_contents
+            )
+        except Exception as e:
+            self.error = f'Could not dispatch match: {e}'
+            return
+
+        # Set the match as pending *after* the RPC call has succeded.
```
Comment (Zopieux) [inline python-lib/prologin/masternode/task.py]: succeeded*

File: python-lib/prologin/masternode/task.py [hunk 8]
```diff
@@ -123,15 +163,42 @@ def __init__(self, config, mid, players, map_contents):
...
+
+    async def redispatch(self):
+        print(f"Redispatching {self}")
+        await self.db.execute(
+            "set_match_status", {"match_status": "new", "match_id": self.mid}
+        )
+
+    async def fail(self):
+        await self.db.execute(
+            "set_match_status", {"match_status": "fail", "match_id": self.mid},
```
Comment (seirl) [inline python-lib/prologin/masternode/task.py:203]: `match_status` should be `failed` here I think?

**Pull request merged**